Pragma bug fix/enhancement

tauren opened the pull request:
This PR includes the following:

* Fix broken build. Previously the build would only work if `babel-cli` and `rimraf` were installed globally.
* Fix broken `pragma` feature. Previously, when a `prama` option was specified, `createFactory` was undefined.
* Add `this.options.imports` feature, enabling user to specify an import statement (or array of imports)

tauren committed:
Make sure build works when babel and rimraf are not globally installed

tauren committed:
Fix bug where `createFactory` is not defined when `pragma` is set.

tauren committed:
Add support for `imports` option

this.options.imports can be a string or an array

tauren committed:
Webpack config example using mdx-loader options

jamesknelson commented:
Thanks for picking up on the broken build!

Re: the missing `createFactory`, my thought was that if you're specifying a `pragma` option, you wouldn't want to use the default `React.createFactory`. Perhaps it would be better to leave the React import out when `pragma` is specified, but document that you'll need to import your own `createFactory`?

I like the `imports` option though. Would you be able to add a test?

tauren committed:
Remove dependency on React when pragma is specified

tauren committed:
Add tests and fix broken tests

tauren committed:
Update mdx-loader to handle pragma in factories

tauren commented:
Yes, you're right. If `pragma` is defined, there should be no assumption that React is available. I've fixed this.

I've also fixed the broken tests and added additional tests for `imports` option.

Lastly, the default `codeBlock` factory assumes React is available as well. I've updated `mdx-loader` to overwrite it appropriately.

I briefly looked over the package structure and It isn't clear how `mdx-loader` gets published separately to npm. Would you mind explaining?

jamesknelson commented:
This looks great, thanks!

mdx-loader is actually a separate package: https://github.com/jamesknelson/mdx-loader

The mdx-loader in this project is for the not-yet-complete website for mdxc. I've inlined it here as I don't want to rely on a separate package. Could you move the changes to mdx-loader to a separate PR on that package, then I'll publish them?

tauren commented:
I've updated the `mdx-loader` repo, but do you really want the changes to mdx-loader removed from this repo? Would love to see these changes published to NPM soon!

jamesknelson commented:
Thanks! I've just merged and published mdxc 1.0.0-beta.6 and mdx-loader 1.0.0-beta.3
